<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>fastStructure by rajanil</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/respond.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!--[if lt IE 8]>
    <link rel="stylesheet" href="stylesheets/ie.css">
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>

  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-57610622-1', 'auto');
    ga('send', 'pageview');
    
    $( document ).ready(function() {
      $('#header .fork').on('click', function() {
        ga('send', 'event', 'button', 'click', 'fork');
      });
      $('#header .zip').on('click', function() {
        ga('send', 'event', 'button', 'click', 'zip');
      });
      $('#header .tar').on('click', function() {
        ga('send', 'event', 'button', 'click', 'tar');
      });
    });
  

  </script>


  <body>
      <div id="header">
        <nav>
          <li class="fork"><a href="https://github.com/rajanil/fastStructure">View On GitHub</a></li>
          <li class="downloads zip"><a href="https://github.com/rajanil/fastStructure/zipball/master">ZIP</a></li>
          <li class="downloads tar"><a href="https://github.com/rajanil/fastStructure/tarball/master">TAR</a></li>
          <li class="title">DOWNLOADS</li>
        </nav>
      </div><!-- end header -->

    <div class="wrapper">

      <section>
        <div id="title">
          <h1>fastStructure</h1>
          <p>A variational framework for inferring population structure from SNP genotype data.</p>
          <hr>
          <span class="credits left">Project maintained by <a href="https://github.com/rajanil">Anil Raj</a></span>
          <span class="credits right">Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></span>
        </div>

        <h1>
<a name="Introduction" class="anchor" href="#Introduction"><span class="octicon octicon-link"></span></a>Introduction</h1>

<p><em>fastStructure</em> is an algorithm for <a href="http://biorxiv.org/content/early/2013/12/02/001073">inferring population structure from large SNP genotype data</a>. It is based on a variational Bayesian framework for posterior inference and is written in Python2.x. Here, we summarize how to setup this software package, compile the C and Cython scripts and run the algorithm on a test simulated genotype dataset.</p>

<h2>
<a name="citation" class="anchor" href="#citation"><span class="octicon octicon-link"></span></a>Citation</h2>

<p> Anil Raj, Matthew Stephens, and Jonathan K. Pritchard. 
<em>fastSTRUCTURE: Variational Inference of Population Structure in Large SNP Data Sets </em>,
(Genetics) June 2014 197:573-589
[<a href="www.genetics.org/content/197/2/573.full">Genetics</a>, 
<a href="biorxiv.org/content/early/2013/12/02/001073">Biorxiv</a>] </p>

<h2>
<a name="dependencies" class="anchor" href="#dependencies"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<p><em>fastStructure</em> depends on </p>

<ul>
<li><a href="http://www.numpy.org/">Numpy</a></li>
<li><a href="http://www.scipy.org/">Scipy</a></li>
<li><a href="http://cython.org/">Cython</a></li>
<li><a href="http://www.gnu.org/software/gsl/">GNU Scientific Library</a></li>
</ul><p>A number of python distributions already have the first three modules packaged in them. It is also
straightforward to install all these dependencies 
 (1) using package managers for MACOSX and several Linux distributions,
 (2) from platform-specific binary packages, and
 (3) directly from source</p>

<p>Detailed instructions for installing these dependencies are provided in the README file
in the fastStructure software bundle.</p>

<h2>
<a name="getting-the-source-code" class="anchor" href="#getting-the-source-code"><span class="octicon octicon-link"></span></a>Getting the source code</h2>

<p>To obtain the source code from github, let us assume you want to clone this repo into a
directory named proj:</p>

<pre><code>mkdir ~/proj
cd ~/proj
git clone https://github.com/rajanil/fastStructure
</code></pre>

<p>To retrieve the latest code updates, you can do the following:</p>

<pre><code>cd ~/proj/fastStructure
git fetch
git merge origin/master
</code></pre>

<p>You can also retrieve the code using wget by doing the following:</p>

<pre><code>wget --no-check-certificate https://github.com/rajanil/fastStructure/archive/master.tar.gz
</code></pre>

<h2>
<a name="building-python-extensions" class="anchor" href="#building-python-extensions"><span class="octicon octicon-link"></span></a>Building Python extensions</h2>

<p>Before building python extensions, it is important to identify the path to the library
files <em>libgsl.so</em> and <em>libgslcblas.so</em>, and header file 
<em>gsl/gsl_sf_psi.h</em> that are part of your GSL installation. For a default 
installation of GSL, the libraries (<em>.so</em> files) are usually found in /usr/local/lib
and the header files (<em>.h</em> files) in /usr/local/include. In this case, you can add these
lines to your <em>.bashrc</em> file on your home directory.

<pre><code>export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib
export CFLAGS="-I/usr/local/include"
export LDFLAGS="-L/usr/local/lib"
</code></pre>

<p>Then, run <code>source ~/.bashrc</code> to set these environment variables.</p>

<p>To build library extensions, you can do the following:</p>

<pre><code>cd ~/proj/fastStructure/vars
python setup.py build_ext --inplace
</code></pre>

<p>To compile the main cython scripts, you can do the following:</p>

<pre><code>cd ~/proj/fastStructure
python setup.py build_ext --inplace
</code></pre>

<p>Each setup will create some .c and .so (shared object) files.
This setup may give some warnings, which are OK. If you get errors that indicate the
build failed, this might be because the wrong compiler is being used or
environment variables (like LD_LIBRARY_PATH) are set incorrectly. To use a specific
gcc compiler, you can do the following:</p>

<pre><code>CC=/path/to/compiler python setup.py build_ext --inplace
</code></pre>

<h2>
<a name="executing-the-code" class="anchor" href="#executing-the-code"><span class="octicon octicon-link"></span></a>Executing the code</h2>

<p>The main script you will need to execute is structure.py. To see command-line 
options that need to be passed to the script, you can do the following:</p>

<pre><code>$ python structure.py

Here is how you can use this script

Usage: python structure.py
     -K &lt;int&gt;   (number of populations)
     --input=&lt;file&gt;   (/path/to/input/file)
     --output=&lt;file&gt;   (/path/to/output/file)
     --tol=&lt;float&gt;   (convergence criterion; default: 10e-6)
     --prior={simple,logistic}   (choice of prior; default: simple)
     --cv=&lt;int&gt;   (number of test sets for cross-validation, 0 implies no CV step; default: 0)
     --format={bed,str} (format of input file; default: bed)
     --full   (to output all variational parameters; optional)
     --seed=&lt;int&gt;   (manually specify seed for random number generator; optional)
</code></pre>


<p>fastStructure performs inference for the simplest, independent-loci, admixture model, with two choices of priors 
that can be specified using the `--prior` flag. Thus, unlike Structure, fastStructure does not require the 
`mainparams' and `extraparam' files. The inference algorithm used by fastStructure is fundamentally different from
that of Structure and requires the setting of far fewer options. All options for fastStructure can be passed 
via the flags listed above.</p>

<h3>
<a name="main-options" class="anchor" href="#main-options"><span class="octicon octicon-link"></span></a>Main options</h3>

<p>The key options to pass to the scripts are the input file, the output file and the number of populations.
Assuming the input file is named genotypes.bed (with corresponding genotypes.fam and genotypes.bim),
the output file is named <i>genotypes_output</i> and the number of populations you would like is 3, 
you can run the algorithm as follows:</p>

<pre><code>python structure.py -K 3 --input=genotypes --output=genotypes_output
</code></pre>

<p>This generates a genotypes_output.3.log file that tracks how the algorithm proceeds, and files
genotypes_output.3.meanQ and genotypes_output.3.meanP containing the posterior mean of
admixture proportions and allele frequencies, respectively. The orders of samples and 
SNPs in the output files
match those in the `.fam` file and `.bim` file, respectively. Note that input file names should
not include suffixes (e.g., .bed) and are relative to the main project directory (unless a full
path is provided).</p>

<h3>
<a name="input-data-format" class="anchor" href="#input-data-format"><span class="octicon octicon-link"></span></a>Input data format</h3>

<p>The current implementation can import data from <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#bed">plink bed</a>
format and the original Structure format. If the data are in plink format, ensure that
bed, bim and fam files for the dataset are all present in the same path.</p>

<p>While the original Structure program allowed for a more flexible input format, fastStructure expects a more
specific `structure input format'. Specifically, rows in the data file correspond to samples, with two rows per sample
(note that only diploids are handled by this software), and columns correspond to SNPs. The first 6 columns
of the file will be ignored; these typically would include IDs, metadata, etc. This software only 
handles bi-allelic loci. The two alleles at each locus can be encoded as desired; however, missing data 
should be encoded as `-9'. 

<h2>
<a name="running-on-test-data" class="anchor" href="#running-on-test-data"><span class="octicon octicon-link"></span></a>Running on test data</h2>

<p>A test simulated dataset is provided in test/testdata.bed with genotypes sampled for
200 individuals at 500 SNP loci. The output files in test/ were generated as follows:</p>

<pre><code>$ python structure.py -K 3 --input=test/testdata --output=testoutput_simple --full --seed=100
$ ls test/testoutput_simple*
test/testoutput_simple.3.log  test/testoutput_simple.3.meanP  test/testoutput_simple.3.meanQ  
test/testoutput_simple.3.varP  test/testoutput_simple.3.varQ

$ python structure.py -K 3 --input=test/testdata --output=testoutput_logistic --full --seed=100 --prior=logistic
$ ls test/testoutput_logistic*
test/testoutput_logistic.3.log    test/testoutput_logistic.3.meanQ  test/testoutput_logistic.3.varQ
test/testoutput_logistic.3.meanP  test/testoutput_logistic.3.varP

$ tail -n 3 test/testoutput_simple.3.log
Marginal Likelihood (avg over genotypes) = -0.9777044544
Total time = 4.7611 seconds
Total iterations = 160
</code></pre>

<p>Executing the code with the provided test data should generate a log file identical to the ones in test/, 
as a final check that the source code has been downloaded and compiled correctly. The algorithm scales 
linearly with number of samples, number of loci and value of K; the expected runtime for a new dataset can be 
computed from the runtime in the above log file.</p>

<h2>
<a name="choosing-model-complexity" class="anchor" href="#choosing-model-complexity"><span class="octicon octicon-link"></span></a>Choosing model complexity</h2>

<p>In order to choose the appropriate number of model components that explain structure in the dataset,
we recommend running the algorithm for multiple choices of K. We have provided a utility tool to parse
through the output of these runs and provide a reasonable range of values for the model complexity 
appropriate for this dataset.</p>

Assuming the algorithm was run on the test dataset for choices of K ranging from 1 to 10, and
the output flag was --output=test/testoutput_simple, you can obtain the model complexity 
by doing the following:</p>

<pre><code>$ python chooseK.py --input=test/testoutput_simple
Model complexity that maximizes marginal likelihood = 2
Model components used to explain structure in data = 4
</code></pre>

<h2>
<a name="visualizing-admixture-proportions" class="anchor" href="#visualizing-admixture-proportions"><span class="octicon octicon-link"></span></a>Visualizing admixture proportions</h2>

<p>In order to visualize the expected admixture proportions inferred by fastStructure, we have
provided a simple tool to generate <a href="https://web.stanford.edu/group/rosenberglab/distruct.html">Distruct</a> 
plots using the mean of the variational
posterior distribution over admixture proportions. The samples in the plot will be grouped
according to population labels inferred by fastStructure. However, if the user would like to
group the samples according to some other categorical label (e.g., geographic location), these labels
can be provided as a separate file using the flag --popfile. The order of labels in this file (one label per row)
should match the order of samples in the input data files.</p>

<pre><code>$ python distruct.py

Here is how you can use this script

Usage: python distruct.py
     -K &lt;int&gt;  (number of populations)
     --input=&lt;file&gt;  (/path/to/input/file; same as output flag passed to structure.py)
     --output=&lt;file&gt;   (/path/to/output/file)
     --popfile=&lt;file&gt;  (file with known categorical labels; optional)
     --title=&lt;figure title&gt;  (a title for the figure; optional)
</code></pre>

Assuming the algorithm was run on the test dataset for K=5, and
the output flag was --output=test/testoutput_simple, you can generate a Distruct plot
by doing the following:</p>

<pre><code>$ python distruct.py -K 5 --input=test/testoutput_simple --output=test/testoutput_simple_distruct.svg
</code></pre>

      </section>

    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>
